pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/io/ConnectorUtils.java

```diff
@@ -142,49 +149,81 @@ public static TreeMap<String, Connector> searchForConnectors(String connectorsDi
         Path path = Paths.get(connectorsDirectory).toAbsolutePath();
         log.info("Searching for connectors in {}", path);
 
-        TreeMap<String, Connector> connectors = new TreeMap<>();
         if (!path.toFile().exists()) {
             log.warn("Connectors archive directory not found");
-            return connectors;
+            return new TreeMap<>();
         }
 
+        List<Path> archives = new ArrayList<>();
         try (DirectoryStream<Path> stream = Files.newDirectoryStream(path, "*.nar")) {
             for (Path archive : stream) {
-                try {
-
-                    NarClassLoader ncl = NarClassLoaderBuilder.builder()
-                            .narFile(new File(archive.toString()))
-                            .extractionDirectory(narExtractionDirectory)
-                            .build();
-
-                    Connector.ConnectorBuilder connectorBuilder = Connector.builder();
-                    ConnectorDefinition cntDef = ConnectorUtils.getConnectorDefinition(ncl);
-                    log.info("Found connector {} from {}", cntDef, archive);
-
-                    connectorBuilder.archivePath(archive);
-                    if (!StringUtils.isEmpty(cntDef.getSourceClass())) {
-                        if (!StringUtils.isEmpty(cntDef.getSourceConfigClass())) {
-                            connectorBuilder.sourceConfigFieldDefinitions(ConnectorUtils
-                                    .getConnectorConfigDefinition(ncl, cntDef.getSourceConfigClass()));
-                        }
-                    }
+                archives.add(archive);
+            }
+        }
+        if (archives.isEmpty()) {
+            return new TreeMap<>();
+        }
 
-                    if (!StringUtils.isEmpty(cntDef.getSinkClass())) {
-                        if (!StringUtils.isEmpty(cntDef.getSinkConfigClass())) {
-                            connectorBuilder.sinkConfigFieldDefinitions(
-                                    ConnectorUtils.getConnectorConfigDefinition(ncl, cntDef.getSinkConfigClass()));
-                        }
-                    }
+        ExecutorService oneTimeExecutor = null;
+        try {
+            int nThreads = Math.min(Runtime.getRuntime().availableProcessors(), archives.size());
+            log.info("Loading {} connector definitions with a thread pool of size {}", archives.size(), nThreads);
+            oneTimeExecutor = Executors.newFixedThreadPool(nThreads,
+                    new ThreadFactoryBuilder().setNameFormat("search-connectors-executor-%d").build());
```

Comment:
Isn't this completely generic? I'm just wondering about the "search" part in the naming.
```suggestion
                    new ThreadFactoryBuilder().setNameFormat("connector-extraction-executor-%d").build());
```

Reply:
You are right, it is more about extraction or unpacking. The function name is `searchForConnectors` but the parallel work is the extraction. Updating.